### ISSUE TITLE
SANITY CHECK FAILED: [false] at `src/rdb_protocol/terms/type_manip.cc:425` (server is buggy). 

### ISSUE THREAD

**Issue by neumino** [rethinkdb/rethinkdb#4242]
It seems harmless (my server doesn't crash) but it's kind of surprising:

```
r.db('reqlitetest').table('reqlitetestchanges').orderBy({index: 'id'}).info()
```

Throws

```
RqlRuntimeError: SANITY CHECK FAILED: [false] at `src/rdb_protocol/terms/type_manip.cc:425` (server is buggy).  Backtrace:
Sun May 17 19:53:59 2015

1: backtrace_t::backtrace_t() at 0x97e280 (rethinkdb)
2: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at 0x97e54e (rethinkdb)
3: ql::runtime_sanity_check_failed(char const*, int, char const*, std::string const&) at 0xac9394 (rethinkdb)
4: ql::info_term_t::val_info(ql::scope_env_t*, scoped_ptr_t<ql::val_t>) const at 0xa92c9c (rethinkdb)
5: ql::info_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at 0xa935f4 (rethinkdb)
6: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at 0xad062c (rethinkdb)
7: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at 0xacd33d (rethinkdb)
8: ql::query_cache_t::ref_t::run(ql::env_t*, Response*) at 0xae80f1 (rethinkdb)
9: ql::query_cache_t::ref_t::fill_response(Response*) at 0xae9a7e (rethinkdb)
10: ql::run(ql::query_id_t&&, ql::protob_t<Query>, Response*, ql::query_cache_t*, signal_t*) at 0xacd6e0 (rethinkdb)
11: rdb_query_server_t::run_query(ql::query_id_t&&, ql::protob_t<Query> const&, Response*, ql::query_cache_t*, signal_t*) at 0xaf2c2f (rethinkdb)
12: query_server_t::handle(http_req_t const&, http_res_t*, signal_t*) at 0x924562 (rethinkdb)
13: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at 0x96437c (rethinkdb)
14: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at 0x96437c (rethinkdb)
15: http_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at 0x975dab (rethinkdb)
16: std::_Function_handler<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&), boost::_bi::bind_t<void, boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<http_server_t*>, boost::arg<1>, boost::_bi::value<auto_drainer_t::lock_t> > > >::_M_invoke(std::_Any_data const&, scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at 0x976686 (rethinkdb)
17: linux_nonthrowing_tcp_listener_t::handle(int) at 0xbc574c (rethinkdb)
18: coro_t::run() at 0xbb13e8 (rethinkdb)
 in:
r.db("reqlitetest").table("reqlitetestchanges").orderBy({index: "id"}).info()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

```
michel@h9:~$ rethinkdb --version
rethinkdb 2.0.1 (GCC 4.9.2)
```

**Comment by neumino** [comment]
Similar error, though it may not be related:

```
r.map(r.db('test').table('test').changes(), r.range(), function(left, right) {
  return left.merge({num: right})
})
```
```
RqlRuntimeError: SANITY CHECK FAILED: [arg_values[i].has()] at `src/rdb_protocol/var_types.cc:69` (server is buggy).  Backtrace:
Mon May 18 17:45:01 2015

1: backtrace_t::backtrace_t() at 0x97e280 (rethinkdb)
2: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at 0x97e54e (rethinkdb)
3: ql::runtime_sanity_check_failed(char const*, int, char const*, std::string const&) at 0xac9394 (rethinkdb)
4: ql::var_scope_t::with_func_arg_list(std::vector<ql::sym_t, std::allocator<ql::sym_t> > const&, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&) const at 0xa1531e (rethinkdb)
5: ql::reql_func_t::call(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&, ql::eval_flags_t) const at 0xa526f3 (rethinkdb)
6: ql::map_datum_stream_t::next_raw_batch(ql::env_t*, ql::batchspec_t const&) at 0xad6698 (rethinkdb)
7: ql::eager_datum_stream_t::next_grouped_batch(ql::env_t*, ql::batchspec_t const&, std::map<ql::datum_t, std::vector<ql::datum_t, std::allocator<ql::datum_t> >, optional_datum_less_t, std::allocator<std::pair<ql::datum_t const, std::vector<ql::datum_t, std::allocator<ql::datum_t> > > > >*) at 0xad48a7 (rethinkdb)
8: ql::eager_datum_stream_t::next_batch_impl(ql::env_t*, ql::batchspec_t const&) at 0xad4e5d (rethinkdb)
9: ql::datum_stream_t::next_batch(ql::env_t*, ql::batchspec_t const&) at 0xad27b2 (rethinkdb)
10: ql::query_cache_t::ref_t::serve(ql::env_t*, Response*) at 0xae76b0 (rethinkdb)
11: ql::query_cache_t::ref_t::fill_response(Response*) at 0xae9ac2 (rethinkdb)
12: ql::run(ql::query_id_t&&, ql::protob_t<Query>, Response*, ql::query_cache_t*, signal_t*) at 0xacd6e0 (rethinkdb)
13: rdb_query_server_t::run_query(ql::query_id_t&&, ql::protob_t<Query> const&, Response*, ql::query_cache_t*, signal_t*) at 0xaf2c2f (rethinkdb)
14: query_server_t::handle(http_req_t const&, http_res_t*, signal_t*) at 0x924562 (rethinkdb)
15: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at 0x96437c (rethinkdb)
16: routing_http_app_t::handle(http_req_t const&, http_res_t*, signal_t*) at 0x96437c (rethinkdb)
17: http_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at 0x975dab (rethinkdb)
18: std::_Function_handler<void (scoped_ptr_t<linux_tcp_conn_descriptor_t>&), boost::_bi::bind_t<void, boost::_mfi::mf2<void, http_server_t, scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t>, boost::_bi::list3<boost::_bi::value<http_server_t*>, boost::arg<1>, boost::_bi::value<auto_drainer_t::lock_t> > > >::_M_invoke(std::_Any_data const&, scoped_ptr_t<linux_tcp_conn_descriptor_t>&) at 0x976686 (rethinkdb)
19: linux_nonthrowing_tcp_listener_t::handle(int) at 0xbc574c (rethinkdb)
20: coro_t::run() at 0xbb13e8 (rethinkdb)
 in:
r.db("test").table("test").changes().map(r.range(), function(var_59, var_60) { return var_59.merge({num: var_60}); })
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by danielmewes** [comment]
Yeah clearly a bug (or two probably).
Pinging @mlucy

**Comment by mlucy** [comment]
Fix for the first bug is up in CR 3015 by @Tryneus .

**Comment by mlucy** [comment]
The second bug doesn't seem to be reproducible in `next`.  @neumino, did it happen every time for you, or only occasionally?

**Comment by neumino** [comment]
@mlucy -- the second bug was happening every time for me.

**Comment by mlucy** [comment]
A fix for the first bug is in `next`; still working on the second one.

**Comment by mlucy** [comment]
Also still need to backport the first fix to 2.0.x.

**Comment by mlucy** [comment]
Found it; the second bug only seems to occur in the data explorer.  A fix is in CR 3030 by @Tryneus .

**Comment by danielmewes** [comment]
@mlucy Did the second bug affect 2.0.x as well? Can you try backporting both of them today so we can include them in 2.0.4?

**Comment by mlucy** [comment]
@danielmewes -- yes, they both affected both versions.  I'll get them into both branches today.

**Comment by mlucy** [comment]
Fixes for both are in both `next` and `v2.0.x`.
